### ISSUE TITLE
This page is irrelvant if Swashbuckle doesn't work

### ISSUE THREAD

**Issue by lukos** [dotnet/AspNetCore.Docs#18818]
This page lists the steps required to get Swashbuckle working correctly and it works locally but when deployed behind IIS it doesn't work, even with the documented relative path in `SwaggerEndpoint`. 

A task is raised against Swashbuckle (1392) but hasn't even been commented on since November 2019 (apart from by me!)

If this cannot be fixed, the documentation either needs to say it cannot be used on IIS or otherwise remove Swashbuckle and only include NSwag. It is shame to have wasted 2 hours and countless deployment builds for something that simply doesn't work!

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 14fce65f-1337-58cb-2193-8d8e0e62792d
* Version Independent ID: 59222460-a971-e76e-4fe9-23d4d7d768ee
* Content: [Get started with Swashbuckle and ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/tutorials/getting-started-with-swashbuckle?view=aspnetcore-3.1&tabs=visual-studio)
* Content Source: [aspnetcore/tutorials/getting-started-with-swashbuckle.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/tutorials/getting-started-with-swashbuckle.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-tutorials**
* GitHub Login: @zuckerthoben
* Microsoft Alias: **scaddie**

**Comment by Rick-Anderson** [comment]
> A task is raised against Swashbuckle (1392) but hasn't even been commented on since November 2019 (apart from by me!)

@lukos  can you provide a link to that?

**Comment by lukos** [comment]
Sure. It is [here](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/1392)

**Comment by scottaddie** [comment]
@lukos Some questions for you:

1. Where are you using IIS (e.g., Azure App Service on Windows, Windows Server, etc.)?
2. Which version of the `Swashbuckle.AspNetCore` package are you using?
3. Which version of .NET Core are you targeting?

**Comment by stevebelskie** [comment]
@scottaddie I'm able to reproduce the issue on IIS running locally on my windows machine. I can reproduce with `Swashbuckle.AspNetCore` on versions 5.0 (what the doc shows) and version 5.5 (the latest). My app targets 3.1. The app is running as a virtual application nested in another site.

I was also able to get it working by using `../swagger/v1/swagger.json` instead of `./swagger/v1/json`. So the the swagger configuration looks like this.
```cs
app.UseSwagger();
app.UseSwaggerUI(c =>
{
      c.SwaggerEndpoint("../swagger/v1/swagger.json", "My API V1");
});
```
This worked for both locally and on the IIS app, with the only drawback being that the swagger UI generates the url to the swagger doc in a rather ugly (though functional!) fashion.
![image](https://user-images.githubusercontent.com/37506132/85794840-40e9b900-b705-11ea-8066-023cdaae3d13.png)

Swashbuckle appears have made some recent changes in bug fix that I believe explain the odd url construction. ([link](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/b7cf75e7905050305b115dd96640ddd6e74c7ac9/src/Swashbuckle.AspNetCore.SwaggerUI/index.html#L51)). 

A workaround I found, that I do not understand even slightly, was to adjust the `RouteTemplate` and shorten the Swagger endpoint url.

```cs
app.UseSwagger(c => c.RouteTemplate = "/swagger/{documentName}/swagger.json"); //by default the url does not have the leading slash:  swagger/{documentName}/swagger.json
app.UseSwaggerUI(c =>
{
    c.SwaggerEndpoint("v1/swagger.json", "My API V1");
});
```

So two workarounds to get the basic example running for an app in a virtual directory. The first feels closer to the correct way to do it, with the url generation issue seeming like a bug on Swashbuckle's part (I may raise an issue there after a bit more investigation). No idea about the second way as I don't understand why making the route template an absolute url would produce this behavior.

**Comment by domaindrivendev** [comment]
@stevebelskie - the following should work without issue in both locally and IIS/virtual dir environments. If you've tried this with the latest version and are still having issues, please create a new issue in the [Swashbuckle.AspNetCore repo](https://github.com/domaindrivendev/Swashbuckle.AspNetCore)

```csharp
app.UseSwagger();
app.UseSwaggerUI(c =>
{
    c.SwaggerEndpoint("v1/swagger.json", "My API V1");
});
```

Re understanding the effect of the "shortened" endpoint string, there's nothing magic here it's just plain old HTTP semantics - specifically the difference between a _relative_ and _absolute_ url path.

The endpoint you specify here is ultimately called via AJAX from the Swagger UI index page and therefore can be _relative_ to that page. The longer endpoint string `/swagger/v1/swagger.json` is an _absolute_ path (due to the leading `/`) and therefore will always be appended to the host portion of the Swagger UI index URL. `v1/swagger.json` on the other hand is a _relative_ path (due to the lack of a leading `/`) and will therefore be appended to everything before `index.html` in the Swagger UI index URL.

In summary, if your Swagger UI page is served at `http://tempuri.org/virtualdir/swagger/index.html`:

If you specify a _absolute_ endpoint path of `/swagger/v1/swagger.json` then the UI will look for the Swagger JSON at:
`http://tempuri.org/swagger/v1/swagger.json`

Alternatively, if you specify a _relative_ endpoint path of `v1/swagger.json` then the UI will look for the Swagger JSON at:
`http://tempuri.org/virtualdir/swagger/v1/swagger.json`

The latter is not a "workaround", it's just smart use of relative URLs to ensure a configuration that works well in local, IIS virtual dir & reverse proxy environments.

**Comment by Rick-Anderson** [comment]
Moved to Master issue: Swashbuckle #21141
